### PR TITLE
Feature/13 5 profiles

### PR DESCRIPTION
**Pull request by ckunki**
support profiles to enable simplified URIs and to provide password

### CODE REVIEW

File: doc/user_guide/user_guide.md [hunk 1]
```diff
@@ -90,47 +90,70 @@ The first path element is always the name of the bucket. The next elements are t
...
+WritePass = <value>
+```
 
-| Parameter                       | Environment variable | Default value |
-|---------------------------------|----------------------|---------------|
-| Host address of BucketFS server | `BUCKETFS_HOST`      | `localhost`   |
-| Port                            | `BUCKETFS_PORT`      | `2580`        |
-| Write passsword                 | `BUCKETFS_PASSWORD`  | (none)        |
-| Name of root bucket             | `BUCKETFS_BUCKET`    | `default`     |
+Additionally the `<value>` is base64 encoded.  For additional convenience BFSC allows you to provide the password in base64 encoded format and let BFSC decode it with commandline flag `--decode-base64-password` or `-d`. This applies to all methods providing the password: via the profile in your configuration file, as well as via interactive prompt.
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:104]: This flag should be specified in the profile, not on the command line.

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -90,47 +90,70 @@ The first path element is always the name of the bucket. The next elements are t
...
+
+The configuration file uses the INI-file syntax. It is devided into sections. Each section contains a number of lines defining a *profile*. The first line of each section specifies the name of the profile in brackets. The default profile's name is `default`. Each of the following lines of the section may assign a value to an *element*.
+
+BFCS uses the following elements of your profile with the specified hard coded default values:
+
+| Parameter                       | Profile Element | Default value |
+|---------------------------------|-----------------|---------------|
+| Host address of BucketFS server | `host`          | `localhost`   |
+| Port                            | `port`          | `2580`        |
+| Name of root bucket             | `bucket`        | (none)        |
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:129]: Default value could be `default`?

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -90,47 +90,70 @@ The first path element is always the name of the bucket. The next elements are t
...
 ```bash
-bfsc cp -d a.txt bfs:/
+bfsc cp foo.jar bfs://1.2.3.4:8888/simba/drivers/foo.jar
 ```
 
+#### Name of the Root Bucket
+
+BFSC cannot detect whether the URL on the command line contains a bucket name or directly starts with the path *inside* the bucket. If the profile in your configuration file specifies a bucket name, then BFSC will always prepend that bucket name to the path of the URL.
+
+If you want to provide the name of the root bucket on the command line then please do not set `bucket` in your profile.
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:155]: 👍

File: doc/user_guide/user_guide.md [hunk 4]
```diff
@@ -90,47 +90,70 @@ The first path element is always the name of the bucket. The next elements are t
...
+BFSC uses the following precedence for URI parts
+1. Supplied on the command line, or entered interactively in case of the [password](#password-protected-bucket-access)
+2. Elements of your profile
+3. Hard coded default value for host and port
+
+You can define the profile in BFSC's configuration file in your home directory: `~/.bucketfs-client/config`. On Windows the home directory is `%USERPROFILE%`.
+
+The configuration file uses the INI-file syntax. It is devided into sections. Each section contains a number of lines defining a *profile*. The first line of each section specifies the name of the profile in brackets. The default profile's name is `default`. Each of the following lines of the section may assign a value to an *element*.
+
+BFCS uses the following elements of your profile with the specified hard coded default values:
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:123]: How can you select a profile on the command line?

**Review by kaklakariada — CHANGES_REQUESTED** [review]
First part of the review

File: src/test/java/com/exasol/bucketfs/client/BFSC.java [hunk 5]
```diff
@@ -31,32 +41,72 @@ public BFSC feedStdIn(final String in) {
...
+
+    BFSC withConfigFile(final Path path) {
+        this.configFile = path;
+        return this;
     }
 
     /**
      * Run the BucketFS client.
+     *
+     * @throws Exception
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/client/BFSC.java:62]:
```suggestion
```
Doesn't throw an exception?

File: src/test/java/com/exasol/bucketfs/client/ListCommandIT.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        final String path = SETUP.getDefaultBucketUriToFile("non-existing-folder/");
+        final BFSC client = BFSC.create("ls", path).catchStdout();
+        assertExitWithStatus(1, () -> client.run());
+    }
+
+    private static void createFiles() throws BucketAccessException {
+        final UnsynchronizedBucket bucket = SETUP.getDefaultBucket();
+        Stream.of("b.txt", "a.txt", "folder/b1.txt", "folder/a1.txt") //
+                .forEach(path -> createFile(bucket, path));
+        System.out.println("Actual content: " + bucket.listContents());
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/client/ListCommandIT.java:78]:
```suggestion
```

File: src/test/java/com/exasol/bucketfs/client/ListCommandIT.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    void root() {
+        verifyListCommand("", List.of("a.txt", "b.txt", "folder/"));
+    }
+
+    @Test
+    void subfolder() {
+        verifyListCommand("folder/", a -> true, List.of("a1.txt", "b1.txt"));
+    }
+
+    // not test for empty folders as these are not possible in BucketFS
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/client/ListCommandIT.java:65]:
```suggestion
    // There is no test for empty folders as these are not possible in BucketFS.
```

File: src/test/java/com/exasol/bucketfs/profile/ProfileTest.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        final Profile testee = Profile.empty();
+        assertThat(testee.host(), nullValue());
+        assertThat(testee.port(), equalTo(BucketFsUrl.UNDEFINED_PORT));
+        assertThat(testee.bucket(), nullValue());
+        assertThat(testee.password(), nullValue());
+    }
+
+    @Test
+    void equalsContract() {
+        EqualsVerifier.simple().forClass(Profile.class).verify();
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/profile/ProfileTest.java:25]: 👍

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/bucketfs/client/ListCommand.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                .forEach(System.out::println); // NOSONAR
+        // The purpose of the list command is to print the listing on stdout
+        return CommandLine.ExitCode.OK;
+    }
+
+    private BucketFsUrl bucketFsUrl() {
+        try {
+            return BucketFsUrl.from(this.uri, this.profileProvider.getProfile());
+        } catch (final MalformedURLException exception) {
+            throw new BucketFsClientException(ExaError.messageBuilder("E-BFSC-5") //
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/bucketfs/client/ListCommand.java:50]: Will be fixed in next PR on branch 13-6-delete.

File: doc/user_guide/user_guide.md (hunk 1 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:104]: OK

File: doc/user_guide/user_guide.md (hunk 4 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:123]: Not yet, see list of features in ticket #13 😃

File: doc/user_guide/user_guide.md (hunk 2 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:129]: Unfortunately not, see section "Name of the Root Bucket".

File: src/main/java/com/exasol/bucketfs/client/CopyCommand.java [hunk 10]
```diff
@@ -71,10 +74,10 @@ private Path convertSpecToPath(final URI pathSpec) {
 
     private BucketFsUrl createDestinationBucketFsUrl() {
         try {
-            return BucketFsUrl.create(this.destination);
+            return BucketFsUrl.from(this.destination, this.profileProvider.getProfile());
         } catch (final MalformedURLException exception) {
             throw new BucketFsClientException(ExaError.messageBuilder("E-BFSC-3")
-                    .message("Illegal BucketFS destination URL: {{url}}", this.destination).toString());
+                    .message("Invalid BucketFS destination URL: {{url}}", this.destination).toString());
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/bucketfs/client/CopyCommand.java:80]: Will be fixed in next PR, see branch 13-6-delete

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**